# Введение в искусственные нейронные сети
# Урок 2. Keras

## Простая нейросеть на Keras

Давайте попрубуем сделать нейросеть на Keras использую полученные выше знания. Попробуем обучить нейросеть различать рукописные цифры.

In [ ]:
# The full neural network code!
###############################
import numpy as np
#import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.datasets import fashion_mnist


In [ ]:
summary=[]

In [ ]:

(train_images,train_labels),(test_images,test_labels)=fashion_mnist.load_data()

#train_images = mnist.train_images()
#train_labels = mnist.train_labels()
#test_images = mnist.test_images()
#test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) 
test_images = (test_images / 255) 

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

In [ ]:

# Build the model.
model = Sequential([
  Dense(128, activation='relu', input_shape=(784,)),
  
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='kl_divergence',
  metrics=['accuracy'],
)

In [ ]:

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=64,
)

# Evaluate the model.
result=model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # 

# Check our predictions against the ground truths.
print(test_labels[:5]) # 

Epoch 1/5
938/938 [==============================] - 3s 3ms/step - loss: 0.5185 - accuracy: 0.8167
Epoch 2/5
938/938 [==============================] - 3s 3ms/step - loss: 0.3740 - accuracy: 0.8642
Epoch 3/5
938/938 [==============================] - 3s 3ms/step - loss: 0.3327 - accuracy: 0.8781
Epoch 4/5
938/938 [==============================] - 3s 3ms/step - loss: 0.3097 - accuracy: 0.8860
Epoch 5/5
313/313 [==============================] - 0s 1ms/step - loss: 0.3420 - accuracy: 0.8751
[9 2 1 1 6]
[9 2 1 1 6]


In [ ]:
summary.append(['Сеть : 3 слоя (128,64,10), loss=kl_divergence, activation=relu, batch_size=16, ',f'Accuracy:  {result[1]}'])

In [ ]:
summary

[['Сеть : 3 слоя (64,64,10), loss=categorical_crossentropy, activation=relu, ',
  'Accuracy:  0.8671000003814697'],
 ['Сеть : 3 слоя (128,64,10), loss=categorical_crossentropy, activation=relu, ',
  'Accuracy:  0.8754000067710876'],
 ['Сеть : 3 слоя (256,64,10), loss=categorical_crossentropy, activation=relu, ',
  'Accuracy:  0.8705000281333923'],
 ['Сеть : 3 слоя (256,64,10), loss=kl_divergence, activation=relu, ',
  'Accuracy:  0.8766999840736389'],
 ['Сеть : 3 слоя (128,64,10), loss=kl_divergence, activation=relu, ',
  'Accuracy:  0.8805000185966492'],
 ['Сеть : 3 слоя (128,64,10), loss=kl_divergence, activation=sigmoid , ',
  'Accuracy:  0.871399998664856'],
 ['Сеть : 4 слоя (256,128,64,10), loss=kl_divergence, activation=sigmoid , ',
  'Accuracy:  0.8715000152587891'],
 ['Сеть : 3 слоя (128,64,10), loss=kl_divergence, activation=relu, batch_size=16, ',
  'Accuracy:  0.8719000220298767'],
 ['Сеть : 3 слоя (128,64,10), loss=kl_divergence, activation=relu, batch_size=16, ',
  'Accura

## Практическое задание

<ol>
    <li>Попробуйте обучить нейронную сеть на Keras на Fashion-MNIST датасете. 
        Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность?</li>
</ol>

Лучший результат accuracy: 0.88. Улучшить результат по отношению к нейросети с урока помогла смена меры ошибки на kl_divergence. Активационная функция прироста не дала, добавление одного слоя нейронов сделало расчеты сложнее, но не дало прироста в качестве. Изменение кол-ва нейронов в 1 слое до 128 дало прирост качества.



## Дополнительные материалы

<ol>
    <li>https://keras.io/</li>
</ol>

## Используемая литература 

Для подготовки данного методического пособия были использованы следующие ресурсы:
<ol>
    <li>https://keras.io/</li>
    <li>Шакла Н. — Машинное обучение и TensorFlow 2019</li>
    <li>Википедия</li>
    
</ol>